sound/soc/sof/ops.c

```diff
@@ -142,37 +142,43 @@ void snd_sof_dsp_update_bits_forced(struct snd_sof_dev *sdev, u32 bar,
 EXPORT_SYMBOL(snd_sof_dsp_update_bits_forced);
 
 int snd_sof_dsp_register_poll(struct snd_sof_dev *sdev, u32 bar, u32 offset,
-			      u32 mask, u32 target, u32 timeout)
+			      u32 mask, u32 target, u32 time)
 {
-	int time, ret;
+	u32 reg;
+	unsigned long timeout;
+	int k = 0, s = 500;
 
 	/*
-	 * we will poll for couple of ms using mdelay, if not successful
-	 * then go to longer sleep using usleep_range
+	 * split the loop into sleeps of varying resolution. more accurately,
+	 * the range of wakeups are:
+	 * Phase 1(first 5ms): min sleep 0.5ms; max sleep 1ms.
+	 * Phase 2:( 5ms to 10ms) : min sleep 0.5ms; max sleep 10ms
+	 * (usleep_range (500, 1000) and usleep_range(5000, 10000) are
+	 * both possible in this phase depending on whether k > 10 or not).
+	 * Phase 3: (beyond 10 ms) min sleep 5ms; max sleep 10ms.
 	 */
 
-	/* check if set state successful */
-	for (time = 5; time > 0; time--) {
-		if ((snd_sof_dsp_read(sdev, bar, offset) & mask) == target)
+	timeout = jiffies + msecs_to_jiffies(time);
+	reg = snd_sof_dsp_read(sdev, bar, offset);
```

Comment:
You could avoid having to repeat the snd_sof_dsp_read() call outside and inside the loop and variable initialisation at the top of the function by doing something like
`
	for (k = 0, s = 500; ; k++) {
		reg = snd_sof_dsp_read(...);
		if (reg & mask == target || time_after(...))
			break;
		...
	}
`

Reply:
> You could avoid having to repeat the snd_sof_dsp_read() call outside and inside the loop and variable initialisation at the top of the function by doing something like
> 
> ```
> 	for (k = 0, s = 500; ; k++) {
> 		reg = snd_sof_dsp_read(...);
> 		if (reg & mask == target)
> 			return 0;
> 		if (time_after(...))
> 			break;
> 		...
> 	}
> ```

I have another solution to avoid the repeat of the snd_sof_dsp_read() call, let me update it in the new PR, @lyakh please help review, thanks.